syntax = "proto3";

option go_package = "xpb/brokers";


service BrokerGatewayService {
  rpc Login(LoginRequest) returns (LoginResponse);
  rpc SendEmailOTP(SendEmailOTPRequest) returns (SendEmailOTPResponse);
  rpc CreateOrder(CreateOrderRequest) returns (CreateOrderResponse);
  rpc CreateFutureOrder(CreateFutureOrderRequest) returns (CreateFutureOrderResponse);
  rpc GetOrderById(GetOrderRequest) returns (GetOrderResponse);
  rpc GetPendingOrders(GetPendingOrdersRequest) returns (GetPendingOrdersResponse);
  rpc GetListFutureOrders(GetListOrdersRequest) returns (GetListOrdersResponse);
  rpc GetOrdersBySession(GetOrdersBySessionRequest) returns (GetOrdersBySessionResponse);
  rpc CancelOrder(CancelOrderRequest) returns (CancelOrderResponse);
  rpc CancelFutureOrder(CancelFutureOrderRequest) returns (CancelOrderResponse);
  rpc UpdateOrder(UpdateOrderRequest) returns (UpdateOrderResponse);
  rpc GetPortfolio(GetPortfolioRequest) returns (GetPortfolioResponse);
  rpc GetAccountBalance(GetAccountBalanceRequest) returns (GetAccountBalanceResponse);
  rpc GetPortfolioByGroupId(GetPortfolioRequest) returns (GetPortfolioResponse);

  // Execution methods
  rpc ExecutionGetPositions(ExecutionGetPositionsRequest) returns (ExecutionGetPositionsResponse);
  rpc ExecutionGetPositionBySymbol(ExecutionGetPositionBySymbolRequest) returns (ExecutionGetPositionBySymbolResponse);
  rpc ExecutionCreateOrder(ExecutionCreateOrderRequest) returns (ExecutionCreateOrderResponse);
  rpc ExecutionCancelOrder(ExecutionCancelOrderRequest) returns (ExecutionCancelOrderResponse);
  rpc ExecutionGetPendingOrders(ExecutionGetPendingOrdersRequest) returns (ExecutionGetPendingOrdersResponse);
  rpc ExecutionGetPendingOrdersBySymbol(ExecutionGetPendingOrdersBySymbolRequest) returns (ExecutionGetPendingOrdersBySymbolResponse);
  rpc ExecutionGetMaxSellQtyBySymbol(ExecutionGetMaxSellQtyBySymbolRequest) returns (ExecutionGetMaxSellQtyBySymbolResponse);
  rpc ExecutionGetMaxSellQtys(ExecutionGetMaxSellQtysRequest) returns (ExecutionGetMaxSellQtysResponse);
}

message GatewayOrderAgg {
  string order_session_id = 1;
  string symbol = 2;
  double price = 3;
  string order_side =4;
  string order_type = 5;
  double quantity = 6;
  double filled_qty = 7;
  double remaining_qty = 8;
  double matched_price = 9;
  optional double avg_price = 10;
  double cash = 11;
  string status = 12; // "pending" | "p_filled" | "filled" | "rejected"
  int64 created_at = 13;
  int64 updated_at = 14;
}
/* OrderInfo - matches xno_execution.gateway.credential_orders table */
message OrderInfo {
  string order_id = 1;
  string user_id = 2;
  string credential_id = 3;
  string symbol = 4;
  string symbol_type = 5;
  string side = 6;
  string order_type = 7;
  double order_price = 8;
  double quantity = 9;
  double filled_qty = 10;
  double remaining_qty = 11;
  string status = 12;
  int64 created_at = 13;
  int64 updated_at = 14;

}

message GetPendingOrderInfo{
  string order_session_id = 1;
  string original_id = 2;
  string symbol = 3;
  double quantity = 4;
  optional double filled = 5;
  optional double remaining = 6;
  int64 created_at =7;
}
message OrderCreatedInfo{
  string symbol = 1;
  string order_session_id = 2;
}

/* Login */
message LoginRequest {
  string credential_id = 1;
  string otp = 2;
}

message LoginResponse {
  bool success = 1;
  string message = 2;
  string access_token =3;
}

/* Create Order - matches models.CreateOrderRequest */
message OrderInputInfo {
  string symbol = 1;
  string symbol_type = 2;
  double price = 3;
  int64 quantity = 4;
  string order_type = 5;
  string order_side = 6;
}

message CreateOrderRequest {
  string access_token = 1;
  string group_id = 2;
  string basket_session_id = 3;
  repeated OrderInputInfo order_info = 4;
}

message CreateOrderResponse {
  bool success = 1;
  string message = 2;
  repeated OrderCreatedInfo orders = 3;
}


/* Get Order - matches models.GetOrderRequest */
message GetOrderRequest {
  string access_token = 1;
  string group_id = 2;
  string order_session_id = 3;
}

message GetOrderResponse {
  bool success = 1;
  string message = 2;
  GatewayOrderAgg order = 3;
}

/* Cancel Order - matches models.CancelOrderRequest */
message CancelOrderRequest {
  string access_token = 1;
  string order_session_id = 3;
}
message CancelFutureOrderRequest{
  string access_token = 1;
  string credential_id = 2;
  string order_session_id = 3;
}

message CancelOrderResponse {
  bool success = 1;
  string message = 2;
}

/* Portfolio */
message PortfolioItem {
  string symbol = 1;
  int64 quantity = 2;
  optional int64 t_0 = 3;
  optional int64 t_1 = 4;
  optional int64 t_2 = 5;
  optional int64 t_3 =6;
  optional int64 max_sell_qty = 7;
  double avg_price = 8;
  double market_price = 9;
}

message GetPortfolioRequest {
  string group_id = 1;
}

message GetPortfolioResponse {
  bool success = 1;
  string message = 2;
  repeated PortfolioItem items = 3;
}

/* Account Balance - matches models.GetAccountBalanceRequest */
message GetAccountBalanceRequest {
  string access_token = 1;
  string group_id = 2;
}

message GetAccountBalanceResponse {
  bool success = 1;
  string message = 2;
  double cash = 3;
  double equity = 4;
}

/* Pending Orders - matches models.GetPendingOrdersRequest */
message GetPendingOrdersRequest {
  string access_token = 1;
  string group_id = 2;
}

message GetPendingOrdersResponse {
  repeated OrderInfo pending_orders = 1;
}

message GetOrdersBySessionRequest{
  string group_id = 1;
  string basket_session_id = 2;
}

message GetOrdersBySessionResponse{
  bool success = 1;
  string message = 2;
  repeated GatewayOrderAgg orders = 3;
}

message GetListOrdersRequest {
  string user_id = 1;
}

message GetListOrdersResponse {
  repeated GatewayOrderAgg orders = 1;
}

/* Update Order - cancel old order, create new order with updated price */
message UpdateOrderRequest {
  string access_token = 1;
  string group_id = 2;
  string order_session_id = 3;
  double price = 4;
}

message UpdateOrderResponse {
  bool success = 1;
  string message = 2;
  string order_session_id = 3;
  OrderInfo order = 4;
}

message CreateFutureOrderRequest {
  string access_token = 1;
  string user_id = 2;
  string symbol = 3;
  string order_side = 4;
  string order_type = 5;
  double price = 6;
  int64 quantity = 7;
}

message CreateFutureOrderResponse {
  bool success = 1;
  string message = 2;
  OrderCreatedInfo order = 4;
}

message ExecutionGetPendingOrdersRequest {
  string access_token = 1;
  string cred_id = 2;
}

message ExecutionGetPendingOrdersResponse {
  repeated GetPendingOrderInfo pending_orders = 1;
}

message ExecutionGetPendingOrdersBySymbolRequest {
  string access_token = 1;
  string cred_id = 2;
  string symbol = 3;
}

message ExecutionGetPendingOrdersBySymbolResponse {
  repeated GetPendingOrderInfo pending_orders = 1;
}

message ExecutionGetPositionsRequest {
  string access_token = 1;
  string cred_id = 2;
}

message PositionInfo {
  string symbol = 1;
  double position = 2;
  optional double sellable = 3;
}

message ExecutionGetPositionsResponse {
  map<string, PositionInfo> positions = 1; // symbol -> PositionInfo
}

message ExecutionGetPositionBySymbolRequest {
  string access_token = 1;
  string cred_id = 2;
  string symbol = 3;
}

message ExecutionGetPositionBySymbolResponse {
  double position = 1;
  optional double sellable = 2;
}

message ExecutionCreateOrderRequest {
  string access_token = 1;
  string cred_id = 2;
  string symbol = 3;
  string order_side = 4;
  string order_type = 5;
  double price = 6;
  double quantity = 7;
}

message ExecutionCreateOrderResponse {
  string original_order_id = 1;
  string order_session_id = 2;
  repeated OrderCreatedInfo orders = 3;

}

message ExecutionCancelOrderRequest {
  string access_token = 1;
  string cred_id = 2;
  string order_session_id = 3;
}

message ExecutionCancelOrderResponse {
  bool success = 1;
  string message = 2;
}

message ExecutionGetMaxSellQtyBySymbolRequest {
  string access_token = 1;
  string cred_id = 2;
  string symbol = 3;
}

message ExecutionGetMaxSellQtyBySymbolResponse {
  double max_sell_qty = 1;
}

message ExecutionGetMaxSellQtysRequest {
  string access_token = 1;
  string cred_id = 2;
}
message MaxSellInfo {
  string symbol = 1;
  optional double max_sell_qtys = 2;
}
message ExecutionGetMaxSellQtysResponse {
  map<string, MaxSellInfo> MaxSellInfos = 1; // symbol -> max_sell_qty
}

// Send email OTP DNSE
message SendEmailOTPRequest{
  string access_token = 1;
  string credential_id = 2;
}

message SendEmailOTPResponse{
  bool success = 1;
  string message = 2;
  string otp = 3;
}